src/value_log.rs

```diff
@@ -0,0 +1,331 @@
+        if opts.in_memory {
+            None
+        } else {
+            Some(Self {
+                core: Arc::new(RwLock::new(Core::new())),
+                dir_path: opts.value_dir.clone(),
+                opts,
+                writeable_log_offset: AtomicU32::new(0),
+            })
+            // TODO: garbage collection
+            // TODO: discard stats
+        }
+    }
+
+    fn file_path(&self, fid: u32) -> PathBuf {
+        vlog_file_path(&self.dir_path, fid)
+    }
+
+    fn populate_files_map(&self) -> Result<()> {
+        // TODO: implement
+        Ok(())
+    }
+
+    fn create_vlog_file(&self) -> Result<(u32, Arc<Mutex<Wal>>)> {
+        let mut core = self.core.write().unwrap();
+        let fid = core.max_fid + 1;
+        let path = self.file_path(fid);
+        let wal = Wal::open(path, self.opts.clone())?;
+        // TODO: only create new files
+        let wal = Arc::new(Mutex::new(wal));
+        assert!(core.files_map.insert(fid, wal.clone()).is_none());
+        assert!(core.max_fid < fid);
+        core.max_fid = fid;
+        // TODO: add vlog header
+        self.writeable_log_offset
+            .store(0, std::sync::atomic::Ordering::SeqCst);
+        core.num_entries_written = 0;
+        Ok((fid, wal))
+    }
+
+    fn sorted_fids(&self) -> Vec<u32> {
+        let core = self.core.read().unwrap();
+        let mut to_be_deleted = HashSet::new();
+        for fid in core.files_to_delete.iter().cloned() {
+            to_be_deleted.insert(fid);
+        }
+        let mut result = vec![];
+        for (fid, _) in core.files_map.iter() {
+            if to_be_deleted.get(fid).is_none() {
+                result.push(*fid);
+            }
+        }
+        result.sort();
+        result
+    }
+
+    /// Open value log directory
+    pub fn open(&self) -> Result<()> {
+        self.populate_files_map()?;
+        if self.core.read().unwrap().files_map.len() == 0 {
+            self.create_vlog_file()?;
+        }
+        // TODO find empty files and iterate vlogs
+        Ok(())
+    }
+
+    fn w_offset(&self) -> u32 {
+        self.writeable_log_offset
+            .load(std::sync::atomic::Ordering::SeqCst)
+    }
+
+    /// write requests to vlog, and put vlog pointers back in `Request`.
+    /// `write` should not be called concurrently, otherwise this will lead to wrong result.
+    pub fn write(&self, requests: &mut [Request]) -> Result<()> {
+        let result = self.write_inner(requests);
+        if self.opts.sync_writes {
+            let core = self.core.read().unwrap();
+            let current_log_id = core.max_fid;
+            let current_log_ptr = core.files_map.get(&current_log_id).unwrap().clone();
+            let mut current_log = current_log_ptr.lock().unwrap();
+            drop(core);
+            current_log.sync()?;
+        }
+        result
+    }
+
+    pub fn write_inner(&self, requests: &mut [Request]) -> Result<()> {
+        // TODO: validate writes
```

Comment:
Seems need to check if it's in memory mode first.

Reply:
If `ValueLog` is created, it must be on-disk. `ValueLog::new` will return `None` if in-memory mode is enabled.